src/optical_spi.c

```diff
@@ -2,143 +2,139 @@
 Optical SPI
 
 This handles the process of communicating with the microcontroller on the
-PAY-Optical board. Using SPI, this microcontroller (PAY) functions as the master while
+PAY-Optical board. Using SPI, this microcontroller (PAY-SSM) functions as the master while
 the PAY-Optical microcontroller functions as the slave.
 
-PAY requests data for a specific field number, then PAY-Optical reads that data
-and sends it back.
+Protocol:
 
+PAY sends 2 bytes:
+byte 1 = command opcode
+byte 2 = well data (or other additional info)
 
-Protocol:
+PAY responds by interpreting the command opcode. OPTICAL performs the command.
+
+Sometimes, PAY's requested command always requires data to be returned (usually 24-bit data from the ADC).
+In this case, OPTICAL loads 1 byte of data into the SPI Data Register (SPDR).
+It will then set its DATA_RDYn line LOW, signaling PAY to initiate a SPI transfer.
+The data between PAY's and OPTICAL's SPDR registers are swapped. This is repeated as many times
+as necessary until all the data is sent from OPTICAL to PAY.
 
-PAY sends 1 byte
-Bits 7-6: if 11, this is a request
-Bits 5-0: the "field number", meaning which sensor to read (currently 0-33, maybe 0-47 later?)
+The SPI Transmission Flag (SPIF) is set at the end of every SPI transmission. It is
+automatically cleared when the SPDR register is accessed.
 
-Repeat 3 times:
-PAY-Optical retrieves the data and sets the DATA_RDY pin high
-PAY interrupt triggered from the DATA_RDY pin
-PAY initiates a 1 byte SPI transfer to get the data
+well_data interpreted as:
+bit 7 - optical density = 0, fluorescent LED = 1
+bit 5-0 = well number (0-31)
 
-PAY now has the 24 bit ADC reading (raw data) from PAY-Optical, which was sent
-as big-endian/network order (most significant byte first)
 
-TODO - maybe put protocol and constants in lib-common?
+TODO:
+- implement timeout
 */
 
 #include "optical_spi.h"
 
-// Must use volatile variables because we modify them in the ISR, but read them
-// from the blocking function
+// volatile needed?
+volatile uint8_t spi_frame_number = 0;  // number of bytes received from OPTICAL
+volatile uint32_t opt_spi_data = 0;     // Data received from OPTICAL, always right-aligned
```

Comment:
You still need some global variable (boolean/flag) to know whether you are currently waiting for a data response, and another global variable to store the field number you are waiting for.

There's no point in making these two variables global because they are only accessed from a piece of code that blocks until its operation is complete.

Also you don't need volatile because nothing is being accessed within an interrupt.